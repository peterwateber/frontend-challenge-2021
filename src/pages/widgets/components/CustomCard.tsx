import Avatar from "@material-ui/core/Avatar"
import Card from "@material-ui/core/Card"
import CardActions from "@material-ui/core/CardActions"
import CardContent from "@material-ui/core/CardContent"
import CardHeader from "@material-ui/core/CardHeader"
import CardMedia from "@material-ui/core/CardMedia"
import Collapse from "@material-ui/core/Collapse"
import { red } from "@material-ui/core/colors"
import IconButton from "@material-ui/core/IconButton"
import { createStyles, makeStyles, Theme } from "@material-ui/core/styles"
import Typography from "@material-ui/core/Typography"
import ExpandMoreIcon from "@material-ui/icons/ExpandMore"
import FavoriteIcon from "@material-ui/icons/Favorite"
import ShareIcon from "@material-ui/icons/Share"
import clsx from "clsx"
import React from "react"
import { useTranslation } from "react-i18next"


interface Props {
}

const CustomCard: React.FC<Props> = (props) => {
    const classes = useStyles()
    const [expanded, setExpanded] = React.useState(false)
    const { t } = useTranslation()

    const handleExpandClick = () => {
        setExpanded(!expanded)
    }

    return (
        <Card className={classes.root}>
            <CardHeader
                avatar={
                    <Avatar aria-label="recipe" className={classes.avatar}>
                        {t("Card.CardName")
                            .split(" ")
                            .map((n) => n[0])
                            .splice(0, 2)
                            .join("")}
                    </Avatar>
                }
                title={t("Card.CardName")}
                subheader="September 14, 2016"
            />
            <CardMedia
                className={classes.media}
                image="https://picsum.photos/400?f"
                title="Paella dish"
            />
            <CardContent>
                <Typography variant="body2" color="textSecondary" component="p">
                    {t("Card.CardDetails")}
                </Typography>
            </CardContent>
            <CardActions disableSpacing>
                <IconButton aria-label="add to favorites">
                    <FavoriteIcon />
                </IconButton>
                <IconButton aria-label="share">
                    <ShareIcon />
                </IconButton>
                <IconButton
                    className={clsx(classes.expand, {
                        [classes.expandOpen]: expanded,
                    })}
                    onClick={handleExpandClick}
                    aria-expanded={expanded}
                    aria-label="show more"
                >
                    <ExpandMoreIcon />
                </IconButton>
            </CardActions>
            <Collapse in={expanded} timeout="auto" unmountOnExit>
                <CardContent>
                    <Typography paragraph>
                        {t("Card.CardCollapseDetails")}
                    </Typography>
                </CardContent>
            </Collapse>
        </Card>
    )
}

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            maxWidth: 345,
        },
        media: {
            height: 0,
            paddingTop: "56.25%", // 16:9
        },
        expand: {
            transform: "rotate(0deg)",
            marginLeft: "auto",
            transition: theme.transitions.create("transform", {
                duration: theme.transitions.duration.shortest,
            }),
        },
        expandOpen: {
            transform: "rotate(180deg)",
        },
        avatar: {
            backgroundColor: red[500],
        },
    })
)

export default CustomCard
