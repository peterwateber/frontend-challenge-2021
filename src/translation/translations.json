{
    "en": {
        "translation": {
            "Overview": "Overview",
            "Card": {
                "CardName": "Peter Wonders",
                "CardDetails": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
                "CardCollapseDetails": "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum."
            },
            "Dialog": {
                "OpenAlertDialog": "Open Alert Dialog",
                "OpenAlertDialogTitle": "Use Google's location service?",
                "OpenAlertDialogContent": "Let Google help apps determine location. This means sending anonymous location data to Google, even when no apps are running.",
                "OpenAlertDialogButton": {
                    "Agree": "Agree",
                    "Disagree": "Disagree"
                }
            }
        }
    },
    "fr": {
        "translation": {
            "Overview": "Overview FR",
            "Card": {
                "CardName": "Peter Wonders FR",
                "CardDetails": "<some translation FR>",
                "CardCollapseDetails": "<some translation FR>"
            },
            "Dialog": {
                "OpenAlertDialog": "Open Alert Dialog FR",
                "OpenAlertDialogTitle": "Use Google's location service? FR",
                "OpenAlertDialogContent": "<some translation FR>",
                "OpenAlertDialogButton": {
                    "Agree": "Agree FR",
                    "Disagree": "Disagree FR"
                }
            }
        }
    },
    "de": {
        "translation": {
            "Overview": "Overview DE",
            "Card": {
                "CardName": "Peter Wonders DE",
                "CardDetails": "<some translation DE>",
                "CardCollapseDetails": "<some translation DE>"
            },
            "Dialog": {
                "OpenAlertDialog": "Open Alert Dialog DE",
                "OpenAlertDialogTitle": "Use Google's location service? DE",
                "OpenAlertDialogContent": "<some translation DE>",
                "OpenAlertDialogButton": {
                    "Agree": "Agree DE",
                    "Disagree": "Disagree DE"
                }
            }
        }
    }
}
