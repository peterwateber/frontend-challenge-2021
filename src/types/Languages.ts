/** Languages */
export enum LanguagesCode {
    English = "en",
    French = "fr",
    German = "de",
}

export const MapLanguageKey = {
    [LanguagesCode.English]: "English",
    [LanguagesCode.French]: "French",
    [LanguagesCode.German]: "German",
}

export const Languages = [
    {
        name: MapLanguageKey[LanguagesCode.English],
        code: LanguagesCode.English,
    },
    {
        name: MapLanguageKey[LanguagesCode.French],
        code: LanguagesCode.French,
    },
    {
        name: MapLanguageKey[LanguagesCode.German],
        code: LanguagesCode.German,
    },
]
