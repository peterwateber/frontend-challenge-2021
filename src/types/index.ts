export * from "./Languages"
export * from "./Store"