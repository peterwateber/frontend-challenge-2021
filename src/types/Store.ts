import { LanguagesCode } from "./Languages"

export enum FrontEndWidgets {
    Card = "Card",
    Dialog = "Dialog",
    Datepicker = "Datepicker",
}

export interface Widget {
    id: string,
    widget: FrontEndWidgets
    language: LanguagesCode
}

export interface WidgetsState {
    widgets: Widget[]
    currentLanguage: LanguagesCode
}

export const localStorageName = "items"
