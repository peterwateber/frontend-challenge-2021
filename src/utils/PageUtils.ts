import React from "react"
import { FrontEndWidgets } from "types"

interface TypeObj {
    [key: string]: any
}

export const sanitizeInput = (obj: TypeObj) => {
    let flag = false,
        errors = ""
    if (!Boolean(Object.values(obj).length)) throw new Error("Fields could not be empty.")
    Object.keys(obj).forEach((input: string) => {
        const _value = (obj[input] || "").toString().trim()
        if (!Boolean(_value)) {
            flag = true
            errors += `Field '${input}' cannot be empty.<br/>`
        }
        obj[input] = _value
    })
    if (flag) {
        throw new Error(errors.trim())
    }
    return obj
}

export const widgetOptions = (widget: FrontEndWidgets, idx: number) => {
    const path =
        widget === FrontEndWidgets.Datepicker
            ? `Form/Datepickers`
            : `Custom${widget}`
    return {
        id: idx,
        name: widget,
        component: React.lazy(
            () => import(`../pages/widgets/components/${path}`)
        ),
    }
}
