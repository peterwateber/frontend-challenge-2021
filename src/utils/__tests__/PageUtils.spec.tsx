import React from "react"
import { FrontEndWidgets } from "types"
import { sanitizeInput, widgetOptions } from "../PageUtils"
import { render } from "@testing-library/react"
import CustomCard from "pages/widgets/components/CustomCard"
import "../../translation.config"

describe("PageUtils", () => {
    it("should throw string error for empty object values for sanitizeInput()", () => {
        expect(() =>
            sanitizeInput({
                title: "",
                message: "",
            })
        ).toThrowError(
            "Field 'title' cannot be empty.<br/>Field 'message' cannot be empty.<br/>"
        )
    })

    it("should load component pased on widgetOptions()", () => {
        const Comp = widgetOptions(FrontEndWidgets.Card, 1)
        render(
            <React.Suspense fallback="loading">
                <Comp.component key="1" />
            </React.Suspense>
        )
        const card = render(
            <React.Suspense fallback="loading">
                <CustomCard />
            </React.Suspense>
        )
        expect(card).toMatchSnapshot()
    })
})
