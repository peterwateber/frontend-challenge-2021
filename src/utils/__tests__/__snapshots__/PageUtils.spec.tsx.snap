// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageUtils should load component pased on widgetOptions() 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      loading
    </div>
    <div>
      <div
        class="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
        data-testid="card"
      >
        <div
          class="MuiCardHeader-root"
        >
          <div
            class="MuiCardHeader-avatar"
          >
            <div
              aria-label="recipe"
              class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-5 MuiAvatar-colorDefault"
            >
              PW
            </div>
          </div>
          <div
            class="MuiCardHeader-content"
          >
            <span
              class="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
            >
              Peter Wonders
            </span>
            <span
              class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
            >
              September 14, 2016
            </span>
          </div>
        </div>
        <div
          class="MuiCardMedia-root makeStyles-media-2"
          style="background-image: url(https://picsum.photos/400?f);"
          title="Paella dish"
        />
        <div
          class="MuiCardContent-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
          >
            Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
          </p>
        </div>
        <div
          class="MuiCardActions-root"
        >
          <button
            aria-label="add to favorites"
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="share"
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81 1.66 0 3-1.34 3-3s-1.34-3-3-3-3 1.34-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9c-1.66 0-3 1.34-3 3s1.34 3 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.16c-.05.21-.08.43-.08.65 0 1.61 1.31 2.92 2.92 2.92 1.61 0 2.92-1.31 2.92-2.92s-1.31-2.92-2.92-2.92z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            aria-expanded="false"
            aria-label="show more"
            class="MuiButtonBase-root MuiIconButton-root makeStyles-expand-3"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
      data-testid="card"
    >
      <div
        class="MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-avatar"
        >
          <div
            aria-label="recipe"
            class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-5 MuiAvatar-colorDefault"
          >
            PW
          </div>
        </div>
        <div
          class="MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
          >
            Peter Wonders
          </span>
          <span
            class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
          >
            September 14, 2016
          </span>
        </div>
      </div>
      <div
        class="MuiCardMedia-root makeStyles-media-2"
        style="background-image: url(https://picsum.photos/400?f);"
        title="Paella dish"
      />
      <div
        class="MuiCardContent-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
        >
          Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
        </p>
      </div>
      <div
        class="MuiCardActions-root"
      >
        <button
          aria-label="add to favorites"
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="share"
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81 1.66 0 3-1.34 3-3s-1.34-3-3-3-3 1.34-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9c-1.66 0-3 1.34-3 3s1.34 3 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.16c-.05.21-.08.43-.08.65 0 1.61 1.31 2.92 2.92 2.92 1.61 0 2.92-1.31 2.92-2.92s-1.31-2.92-2.92-2.92z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-expanded="false"
          aria-label="show more"
          class="MuiButtonBase-root MuiIconButton-root makeStyles-expand-3"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
